src/core/service.c

```diff
@@ -2144,6 +2145,54 @@ _pure_ static bool service_can_reload(Unit *u) {
         return !!s->exec_command[SERVICE_EXEC_RELOAD];
 }
 
+static int service_exec_command_index(Unit *u, ServiceExecCommand id, bool control) {
+        Service *s = SERVICE(u);
+        int index = 0;
+        ExecCommand *current, *first, *c;
+
+        assert(s);
+
+        first = s->exec_command[id];
+        current = control ? s->control_command : s->main_command;
+
+        /* Figure out where we are in the list be walking back to the beginning */
+        for (c = current; c != first; c = c->command_prev)
+                index++;
+
+        return index;
+}
+
+static int service_serialize_exec_command(Unit *u, FILE *f, bool control) {
+        Service *s = SERVICE(u);
+        ServiceExecCommand id;
+        ExecCommand *c;
+        int index;
+        const char *type;
+        _cleanup_free_ char *args = NULL;
+
+        assert(s);
+        assert(f);
+
+        if (control) {
+                type = "control";
+                c = s->control_command;
+                id = s->control_command_id;
+        } else {
+                type = "main";
+                c = s->main_command;
+                id = SERVICE_EXEC_START;
+        }
+
+        assert(c);
+
+        index = service_exec_command_index(u, id, control);
+        args = strv_join(c->argv, " ");
```

Comment:
(or maybe it's even OK if you ignore the OOM condition, as there's little you could do about it. But if you ignore it, then it shouldn't be used as-is in the fprintf() invocation below. Consider simply wrapping it in strempty() there, so that even if this fails we enter a defined state...